Deploy-CAM.ps1

```diff
@@ -2924,58 +2924,101 @@ else {
         }
+
     } while (-not $domainName)
 
+    # Username
+    $username = $null
+    $firstRun = 1
     do {
-        if ( -not $domainAdminCredential ) {
+        if ($domainAdminCredential -and $firstRun) {
+            $username = $domainAdminCredential.UserName
+            $firstRun = $null
+        }
+
+        if ( -not $username) {
             if( -not $deployOverDC ) {
                 $domainAdminMessage = "Please enter the new domain administrator username for the new domain being created"
             }
             else {
                 $domainAdminMessage = "Please enter the username of the service account for Cloud Access Manager to use with the connected domain"
             }
+            $username = Read-Host $domainAdminMessage
+        }
 
-            $domainAdminCredential = Get-Credential -Message $domainAdminMessage
-            $confirmedPassword = Read-Host -AsSecureString "Please re-enter the password"
-
-            # Need plaintext password to check if same
-            $clearPassword = ConvertTo-Plaintext $confirmedPassword
-            if (-not ($domainAdminCredential.GetNetworkCredential().Password -ceq $clearPassword)) {
-                # don't match- try again.
-                Write-Host "The entered passwords do not match."
-                $domainAdminCredential = $null
+        # only check if it is not deployOverDC
+        if (-not $deployOverDC) {
+            if ((-not ($username -match '^[A-Za-z0-9]+(?:[_-][A-Za-z0-9]+)*$')) -or ($username.Length -gt 20)) {
+                Write-Host "`nPlease enter a valid username. It can only contain letters and numbers and cannot be longer than 20 characters." -ForegroundColor Red -BackgroundColor Yellow
+                $username = $null
                 continue
             }
+            # reserved usernames
+            # https://docs.microsoft.com/en-us/azure/virtual-machines/windows/faq
+            $reservedUsername = @('administrator', 'admin', 'user', 'user1', 'test', 'user2', 'test1', 'user3', 'admin1', '1', '123', 'a',
+                                  'actuser', 'adm', 'admin2', 'aspnet', 'backup', 'console', 'david', 'guest', 'john', 'owner', 'root', 'server',
+                                  'sql', 'support', 'support_388945a0', 'sys', 'test2', 'test3', 'user4', 'user5' )
+            
+            if ($username -in $reservedUsername) {
+                Write-Host "`n$username is a reserved username. Please try again" -ForegroundColor Red -BackgroundColor Yellow
+                $username = $null
+                continue                
+            }
         }
 
-        if ((-not ($domainAdminCredential.UserName -imatch '\w+')) -or ($domainAdminCredential.Username.Length -gt 20)) {
-            Write-Host "Please enter a valid username. It can only contain letters and numbers and cannot be longer than 20 characters."
-            $domainAdminCredential = $null
-            continue
+    } while ( -not $username )
+
+    # Password
```

Comment:
This feels like it could be a lot simpler if we set both $password and $confirmedPassword to  $domainAdminCredential.GetNetworkCredential().Password outside of the do loop and then only prompt if not $null (and set both to $null on failure) then don't need the state variables.

Reply:
humm...I could not directly set variable to $domainAdminCredential since it is System.Management.Automation.PSCredential object. That's why at line 3021 I create a new credential object